strawberry/extensions/field_extension.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Awaitable, Callable
+
+from strawberry.utils.cached_property import cached_property
+
+if TYPE_CHECKING:
+    from strawberry.field import StrawberryField
+    from strawberry.types import Info
+
+
+class FieldExtension:
+    def apply(self, field: StrawberryField) -> None:  # pragma: no cover
+        pass
+
+    def resolve(
+        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
+    ) -> Any:  # pragma: no cover
+        raise NotImplementedError(
+            "Sync Resolve is not supported for this Field Extension"
+        )
+
+    async def resolve_async(
+        self, next: Callable[..., Awaitable[Any]], source: Any, info: Info, **kwargs
+    ) -> Any:  # pragma: no cover
+        raise NotImplementedError(
+            "Async Resolve is not supported for this Field Extension"
+        )
+
+    @cached_property
+    def supports_sync(self) -> bool:
+        return type(self).resolve is not FieldExtension.resolve
+
+    @cached_property
+    def supports_async(self) -> bool:
+        return type(self).resolve_async is not FieldExtension.resolve_async
+
+
+class SyncToAsyncExtension(FieldExtension):
+    """Helper class for mixing async extensions with sync resolvers.
+    Applied automatically"""
+
+    async def resolve_async(
+        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
+    ) -> Any:
+        return next(source, info, **kwargs)
+
+
+def ensure_field_extension_compatibility(field: StrawberryField) -> bool:
+    """
+    Verifies that all of the field extensions for a given field support
+    sync or async depending on the field resolver.
+    Inserts a SyncToAsyncExtension to be able to
+    use Async extensions on sync resolvers
+    Throws a TypeError otherwise.
+
+    Returns True if resolving should be async, False on sync resolving
+    based on the resolver and extensions
+    """
+    if not field.extensions or not len(field.extensions):
+        return field.is_async  # pragma: no cover
+
```

Comment:
I think we are not adding the `SyncToAsync` compatibility between all extensions that require one. What do you think about doing something like this to support that:

```python
supports_sync = [extension.supports_sync for extension in field.extensions] + [not field.is_async]
supports_async = [extension.supports_async for extension in field.extensions] + [field.is_async]

previous: Optional[Tuple[bool, bool]] = None
for i, supports_sync, supports_async in enumerate(zip(supports_sync, supports_async)):
    if previous is not None:
        previous_supports_sync, previous_supports_async = previous

        # When the field is async, all extensions must support async as well
        if field.is_async and not supports_async:
            non_async_extensions = ', '.join(
                e.__name__ for e in field.extensions if not e.supports_async
            )
            raise TypeError(
                f"Cannot add sync-only extension(s) {non_async_extension_names} "
                f"to the async resolver of Field {field.name}. "
                f"Please add a resolve_async method to the extension(s)."
            )

        # When previous is sync only and the current extension is async only,
        # there's no way 
        previous_sync_only = previous_supports_sync and not previous_supports_async
        async_only = supports_async and not supports_sync
        if previous_sync_only and async_only:
            # If this was the field itself, we would have raised TypeError
            # in the condition above.
            if i == len(field.extensions):  # pragma: no cover
                raise AssertionError("This should never happen.")

            extension = field.extensions[i]
            previous_extension = field.extensions[i - 1]
            raise TypeError(
                f"sync-only extension {previous_extension.__name__} followed by "
                f"async-only extension {extension.__name__}."
            )

        if previous_supports_async and not supports_async:
            field.extensions.insert(i - 1, SyncToAsyncExtension())


    previous = (supports_sync, supports_async)
```

This should cover all cases, except one: You have an extension that supports both sync and async, followed by a sync only in a sync field. For `schema.execute` it would work correctly, but `schema.execute_sync` would give an error because `SyncToAsync` doesn't have a `resolve` method. This can be fixed by adding one there:

```python
class SyncToAsyncExtension(FieldExtension):
    """Helper class for mixing async extensions with sync resolvers.
    Applied automatically"""

    def resolve(
        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
    ) -> Any:
        return next(source, info, **kwargs)

    async def resolve_async(
        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
    ) -> Any:
        return next(source, info, **kwargs)
```

Reply:
Having the ability to run sync-only extensions on a sync field, then followed by async only extensions is useful! I was worried it was going to add too much complexity, but this seems great, open to adapting it.

I don't exactly understand the case to add `resolve` to the `SyncToAsyncExtension`. This is only used to convert the sync result of a resolver or a sync extension to an `Awaitable`, so that the following `resolve_async` gets the result it expects. All of this is computed before execution of a request, during schema conversion, so we have a static contract of this only being used while converting a sync result to an async.

Reply:
First point resolved in latest commit ☺️